nemo_aligner/algorithms/spin.py

```diff
@@ -0,0 +1,504 @@
+        if max_response_length != response_tokens.size(1):
+            # This may actually happen because NeMo does not always stop generation after `max_length` in batch mode
+            # => `response_tokens` may contain up to `max_length + max_context_length` tokens.
+            # TODO once NeMo fixes this issue we should be able to always raise an exception when the check above fails,
+            # and remove the `if` below.
+            if (
+                max_response_length >= response_tokens.size(1)
+                or response_tokens.size(1) != prompt_lengths.max().item() + self.max_gen_seq_len
+            ):
+                raise AssertionError(
+                    f"max response length ({max_response_length}) does not match the size of "
+                    f"`response_tokens` ({response_tokens.size(1)})"
+                )
+
+        self.model.finish_inference()
+
+        return response_tokens, response_lengths
+
+    def fit(self):
+        if (not isinstance(self.train_dataloader.batch_sampler, MegatronPretrainingRandomBatchSampler)) and (
+            self.cfg.max_epochs is not None and self.cfg.max_epochs > 1
+        ):
+            # if you use MegatronPretrainingBatchSampler as the batch_sampler passed to your train dataloader (in builders.py)
+            # then each epoch will repeat all your samples in the same order as the previous epoch, there is no shuffling
+            # to fix this, you should use MegatronPretrainingRandomBatchSampler instead, which alleviates this issue and allows
+            # random shuffling for each epoch.
+            raise ValueError(
+                "max_epochs > 1 is not supported unless using `MegatronPretrainingRandomBatchSampler` as the batch_sampler for your train dataloader"
+            )
+
+        self.run_timer.start_time()
+
+        iterations_iter = range(self.iteration, self.cfg.max_iterations)
+        if len(iterations_iter) <= 0:
+            # iteration done
+            return
+
+        for _ in iterations_iter:
+            epoch_iter = range(self.epoch, self.cfg.max_epochs)
+            if len(epoch_iter) <= 0:
+                # epoch done
+                return
+
+            for _ in epoch_iter:
+                num_steps_in_epoch = min(
+                    self.max_steps - self.step, self.num_steps_per_epoch - self.step % self.num_steps_per_epoch
+                )
+                loop_iter = range(num_steps_in_epoch)
+
+                if not loop_iter:
+                    return  # training ended
+
+                global_pbar = tqdm(
+                    self.augment_dataloader(self.train_dataloader),
+                    initial=self.step,
+                    total=self.max_steps,
+                    leave=True,
+                    desc="Training steps",
+                )
+
+                for _, global_batch in zip(loop_iter, global_pbar):
+                    self.model.prepare_for_training()
```

Comment:
No, it is necessary because doing generation messes up the various MBS/GBS global vars and they need to be reset again. You get crashes if you remove that line

Reply:
RIght but then can you do it just once after generation instead of on each batch?

Reply:
No, because look at how `global_pbar` is constructed:

```
global_pbar = tqdm(
                    self.augment_dataloader(self.train_dataloader),
                    initial=self.step,
                    total=self.max_steps,
                    leave=True,
                    desc="Training steps",
                )
```

We wrap the train_dataloader with the augment function, which means for every iteration of the for-loop, `augment_dataloader` will fire, and `augment_dataloader` includes the generation which resets the global MBS/GBS values, hence we need to reset them after the for-loop fires but before `train_single_step`

Reply:
Got it, thanks, I forgot generation was happening on each batch.